### PR TITLE
Fix new metadata bugs (towards #1618)

### PR DESCRIPTION
**Pull request by rupertford**
The new metadata classes were overly restrictive in the following two cases:

1. scalar arg metadata accesses were only allowed to be gh_read, however they can be gh_sum when iterating over dofs. Note, the constraint that it can't be gh_sum if not iterating over dofs is not implemented yet, see issue #1953.
2. function space metadata which specifies if a function space requires basis or differential basis functions was not allowing any of the different flavours of any_space function spaces.

### CODE REVIEW

**Comment by rupertford** [comment]
Note, this PR also fixes an unrelated error that has crept onto master.

**Comment by rupertford** [comment]
Ready for review from @arporter or @sergisiso

File: src/psyclone/tests/domain/lfric/kernel/meta_funcs_arg_metadata_test.py [hunk 1]
```diff
@@ -183,12 +183,24 @@ def test_function_space_setter_getter():
     funcs_arg = MetaFuncsArgMetadata("w0", basis_function=True)
     with pytest.raises(ValueError) as info:
         funcs_arg.function_space = "invalid"
+    print(str(info.value))
```
Comment (arporter) [inline src/psyclone/tests/domain/lfric/kernel/meta_funcs_arg_metadata_test.py:186]: Please rm this line.

**Review by arporter — APPROVED** [review]
All looks good barring one extra print statement that I shall remove.
All tests and examples are OK with compilation.
Ref. guide builds with no new warnings.
No other doc updates are required.
This is a 'towards' PR so no need to worry about issue number.

File: src/psyclone/tests/domain/lfric/kernel/meta_funcs_arg_metadata_test.py (hunk 1 above)
Comment (arporter) [reply src/psyclone/tests/domain/lfric/kernel/meta_funcs_arg_metadata_test.py]: I've taken the liberty of doing this.

**Pull request merged**